### PR TITLE
[Receipts] Use RTesseract to extract textual content from receipt images

### PR DESCRIPTION
**Pull request by sampoder**
## Summary of the problem

We're about to introduce a bunch exciting features related to a receipt's textual content. See https://github.com/hackclub/hcb/pull/5295. It's all very exciting..

However, our current implementation of extracting a receipt's textual content only supports PDFs. This PR is an experiment into adding OCR via https://github.com/dannnylo/rtesseract so that we can extract text from images.

## Tests

Receipt:

IMG_5625

Text:

```
Store 1507 Dir Mia

Hal
2020 Market Street

SAN FRANCISCO CA 941

iu

00150700500192402

n:(415) 861-7660 Rx (416) 4

SAFEWAY 6).

36-9032

14

rm

7900024500  2@ NESQUIK CHOCOLATE 7.98 7.98 $
3400000966 22 HRSHY SYRUP CARAHE 10.98 10.98 §
3400047056 38 REESES HINIS 17.97, 14.87 §
Wenber Savings -3.00
4900057979 2 MBH’S HLK CHOC 13.08 9.98 $
Wenber Savings ~4.00
4900058098 MBM’S WILK CHOC 6.99 4.99 §
Wenber Savings -2.00
869 42 OREO SANDWICH 27.16 , 13.96 $
Wenber Savings -13.20
GROC MOREDIBLE
1370026358 88 HEFTY DLXCUP26CT 47.90) )3)'92'T
Nenber Savings ~18.00
(G/FROZEN
32 REFRESHE ICE PREM 14.97 9014.97 B
OP NT ALMNDHLK VAN 3.49 3.49 §
58 LUC FRZN DSRT VAN 99.95 | 97.45 §
Merber Savings ~2.50
22 LUC FRZN DSRT.NEAP 15.985 914.98-6
Nenber Savings -1.00
UC FROZEN DESSERT 23.97 22.47 §
Neaber Savings -1.50
7968318005 28 V CRNR 1% HILK 9.98 9.98 §
7969916186 8 OPN NAT OAT DESSRT 17.87 12.37 §
GEN MERCHANDISE
111712083 MR OLEAN NITRILE 17.99 17,99 7
4700096256 FBRZ AIR LINEN SKY 4:99 4.49
Wenber Savings -0.50
3700097795 FBRZ AIR BORA BORA 7.99 79g 1
7675920332 GD CK TRIGGER 16.99... 16,99 7
7675320432 GC ICE CREAM SCOOP 13.99 6.99 |
Menber Savings ~7.00
AKED GOODS
7430004106 48 LD COSKIC BROWNIES 118 1.16 §
PRODUCE
64312604011 BANANAS GREEN 2.66 2.66 §
WT 2.99 Ib @ $0.69 /Ib
64912604011 BANANAS GREEN 2,84 2.84 §
wT 2.97 Ib @ $0.89 /Ib
CABN2GOAOTT | BANANAS GREEN 2.84 2.64 ¢
WI 2.97 Ib @ £0.89 /Ib
WISCELLANEOUS
22a 108 RCVCBLE BAG CHARGE 2 BO crn 2. 50.
TAK 8.74
wan BALANCE 300, 2g
Credit Purchase 02/09/24 14:13
GARD W Heath eRK HH RRB IS:
REP) 491346461930 AUTH) 00591338
PAYMENT AMOUNT 300.28
AL VISA CREDIT
ALD ADDODOODOSTOTO
TVR 0000090000
TSI 0000
Visa 300.28
CHANGE 0,00
YOUR SAVINGS
Herber Savings 60 70
```

Receipt: 

IMG_5846 (1)

Text:

```
a '

?0DUCE
1

SAFEWAY (9

Store 909 Dir Kevin

Main:(415) 665-4136 Rx: 415) 665~0119
730 Taraval St

SAN FRANCISCO CA 94116

00 9

0009091 0724

'G/FROZEN Price
247700 CLOVER CONVENTIONA 7.99
Menber Savings ~1.00

BANANAS YELLOW 1.98
2.23 Ib @ $0.89 /1b

TAX
*#** BALANCE

Credit Purchase 03/09/24 09:01
CARD # ¥xxHMKKHKKERT7193
REF: 960117471980 AUTH: 0003594D

PAYMENT AMOUNT 8.97
A. VISA CREDIT
AID 0000000031010
TVR 0000000000

TSI 0000
Visa 8.97
CHANGE 0.00
YOUR SAVINGS
jember Savings 1.00
lotal 1,00

Total Savinas Value

YOUR REWARDS
Paints Earned Today &
REWARDS AVAILABLE 4
TOTAL NUMBER OF ITEMS SOLD = 2
93/09/24 09:01 909 176 7 8822

Thank you for shopping Safeway|
om Bar euay, FOR U questions call
6-9637 a, Sefeway.com/fary

7 a — = ———
```

Receipt (this one was purposefully tough!):

IMG_5650

Text:

```
BEVERAGE

suB TOTAL

TAX

TOTAL
```

Receipt:

IMG_5849

Text:

```
LA TORTILLA
495 CASTRO ST INIT B
SAN FRANCISCO, CA 94114
4158613990

Cashier: Al

Transaction 900082

Total $62.63 Kad
VISA 7198 VARWY ha

\$ OF

40.63.

eh eS

Sigvature

| agree 1o pay ine ar0ve amount
per tne cardholoer and/or
merchant agreement

Tip Amount Total
sy 39 $72.12

0% 642.53 $75.16
```

It seems to do pretty well with printed images, which most receipts are, and I think GPT-4 should be able to work from this. One concern I had was speed but I couldn't notice too much of a speed difference locally.

Before merging, we'll need to add the buildpack: https://github.com/pathwaysmedical/heroku-buildpack-tesseract

### CODE REVIEW

**Comment by sampoder** [comment]
Tested the La Tortilla one on ChatGPT, it recognised the merchant:

Screenshot 2024-04-17 at 11 42 49 PM

Screenshot 2024-04-17 at 11 42 43 PM

**Comment by sampoder** [comment]
And then I tested the complex Safeway one with Alex, I surprised it got the balance perfect:

Screenshot 2024-04-17 at 11 43 31 PM


Screenshot 2024-04-17 at 11 44 11 PM

Screenshot 2024-04-17 at 11 43 37 PM

File: app/models/receipt.rb [hunk 1]
```diff
@@ -95,8 +95,13 @@ def extract_textual_content
            when "application/pdf"
              pdf_text
            else
-             # Unable to extract text from this file type
-             return nil
+             begin
```
Comment (garyhtou) [inline app/models/receipt.rb]: does rtesseract support all file formats? or should we add a new `when` case here?

File: Gemfile [hunk 2]
```diff
@@ -212,3 +212,5 @@ gem "ledgerjournal"
 gem "doorkeeper", "~> 5.6"
 
 gem "cssbundling-rails", "~> 1.4"
+
+gem "rtesseract"
```
Comment (garyhtou) [inline Gemfile]: does this come with the tesseract binary? or does that need to be installed separately?

Comment (garyhtou) [reply Gemfile]: just a heads up, the binary could significantly increase our heroku slug size

Comment (sampoder) [reply Gemfile]: yep, we'll have to add the buildpack linked in the PR

Comment (garyhtou) [reply Gemfile]: oops my bad. i didn't see that. thanks!

are there any special steps for people using this locally?

File: app/models/receipt.rb (hunk 1 above)
Comment (sampoder) [reply app/models/receipt.rb]: fixed!

File: Gemfile (hunk 2 above)
Comment (sampoder) [reply Gemfile]: I ran `apt install tesseract-ocr` in my Docker container to make it work

Comment (garyhtou) [reply Gemfile]: Should it be added to the aptfile then?

Comment (sampoder) [reply Gemfile]: Just added!

**Comment by sampoder** [comment]
Merged into https://github.com/hackclub/hcb/pull/5295, see above for more context.

**Comment by YodaLightsabr** [comment]
Re-opening this as part of splitting up #5295

**Comment by sampoder** [comment]
Closing in favour of https://github.com/hackclub/hcb/pull/6300:

Screenshot 2024-05-15 at 6 49 48 PM

**Pull request closed without merging**